[exporter/elasticsearch] Add initial support for metrics

andrzej-stencel opened the pull request:
**Description:**

Adds initial support for metrics. Currently only Sum ang Gauge metric types are supported, other metric types are ignored.

**Testing:**

Added unit tests.

**Documentation:**

Added docs in component README.

andrzej-stencel committed:
[exporter/elasticsearch] add metrics support

andrzej-stencel commented on exporter/elasticsearchexporter/attribute.go:
```diff
@@ -37,3 +38,12 @@ func getFromAttributes(name string, resource, scope, record attrGetter) string {
 	return str
 }
+
+func getFromAttributesNew(name string, defaultValue string, attributeMaps ...pcommon.Map) string {
```
I want to create a separate pull request where I replace the deprecated `getFromAttributes` function with the new one.

axw commented on exporter/elasticsearchexporter/exporter.go:
```diff
@@ -132,11 +133,69 @@ func (e *elasticsearchExporter) pushLogRecord(ctx context.Context, resource pcom
+	for i := 0; i < rls.Len(); i++ {
+		rl := rls.At(i)
+		resource := rl.Resource()
+		ills := rl.ScopeMetrics()
```
nit: not clear what "ills" is (is this copied from logs? instruction scope logs?), could we give it a more meaningful name such as "scopeMetrics"?

andrzej-stencel replied:
Good point, I actually hate it when variable names don't make it easy to understand the code. Guilty as charged!
Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33513/commits/1136e74c2d4a675f42c24b2b65c09873367711d4.

axw commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -155,6 +161,131 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
 }
 
+func (m *encodeModel) encodeMetrics(resource pcommon.Resource, metrics pmetric.MetricSlice, _ pcommon.InstrumentationScope) ([][]byte, error) {
+	hasher := sha256.New()
```
Is using a cryptographic hash necessary? I think we should probably use a fast, non-cryptographic hash algorithm like xxhash (specifically github.com/cespare/xxhash/v2)

andrzej-stencel replied:
Very good point. This reminds me that [the OTel Collector contributing guide](https://github.com/open-telemetry/opentelemetry-collector/blob/v0.102.1/CONTRIBUTING.md#recommended-libraries--defaults) specifically recommends [hash/fnv](https://pkg.go.dev/hash/fnv).

andrzej-stencel replied:
Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33513/commits/cf4d392c28e6f67ba9cbee80fa211f887ff11a11.

axw commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -155,6 +161,131 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
+		err := doc.Serialize(&buf, m.dedot)
+
+		if err != nil {
+			fmt.Printf("Serialize error, dropping doc: %v\n", err)
```
Should we be returning the error?

andrzej-stencel replied:
I changed this code to return the error: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33513/commits/2e9cfb82aef46db9d5bbaf6c514e0cca39b2fc00.

axw commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -155,6 +161,131 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
+
+	hasher.Reset()
+
+	hasher.Write([]byte(timestamp.AsTime().Format(time.RFC3339Nano)))
```
Encoding the uint64 bytes should avoid some allocs, and generally be cheaper:

```go
var buf [8]byte
binary.LittleEndian.PutUint64(buf[:], uint64(timestamp))
hash.Write(buf[:])
```

andrzej-stencel replied:
Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33513/commits/cf4d392c28e6f67ba9cbee80fa211f887ff11a11.

axw commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -155,6 +161,131 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
+}
+
+func mapHash(hasher hash.Hash, m pcommon.Map) {
+	// TODO: Only hasing values, not keys? :thinking:
+	m.Range(func(_ string, v pcommon.Value) bool {
+		hasher.Write([]byte(v.Str()))
+		valueHash(hasher, v)
```
Yes, I do think we want to hash the keys as well.
```suggestion
	m.Range(func(k string, v pcommon.Value) bool {
		hasher.Write([]byte(k.Str()))
		valueHash(hasher, v)
```

andrzej-stencel replied:
Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33513/commits/cf4d392c28e6f67ba9cbee80fa211f887ff11a11.

axw reviewed (commented):
Unsolicited review, hope you don't mind :) Looks great, just a few minor comments.

andrzej-stencel committed:
Merge remote-tracking branch 'upstream/main' into add-metrics-to-elasticsearch-exporter

andrzej-stencel committed:
Merge remote-tracking branch 'upstream/main' into add-metrics-to-elasticsearch-exporter

andrzej-stencel committed:
refactor: fix variable names

andrzej-stencel committed:
refactor: replace SHA256 with FNV for hashing

We don't want cryptographic hashing, we want fast hashing.
Using hash/fnv as recommended in [contribution docs](https://github.com/open-telemetry/opentelemetry-collector/blob/v0.102.1/CONTRIBUTING.md#recommended-libraries--defaults).

Also changed to hash by uint32 instead of string because I believe that is sufficient.

Also added hashing of attribute keys, which was missing.

Changed BigEndian to LittleEndian for consistency, although it doesn't really matter.

andrzej-stencel committed:
fix: fail on serialization errors

When any of the documents created for a metrics slice cannot be serialized,
reject the whole metrics slice.

andrzej-stencel commented:
> Unsolicited review, hope you don't mind :) Looks great, just a few minor comments.

Thank you so much @axw for the review, your comments were very helpful! Keep them coming! :pray:

axw commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -155,6 +161,126 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
+		valueHash(hasher, v)
+
+		return true
+	})
+}
+
+func valueHash(h hash.Hash, v pcommon.Value) {
+	switch v.Type() {
+	case pcommon.ValueTypeEmpty:
+		h.Write([]byte{0})
+	case pcommon.ValueTypeStr:
+		h.Write([]byte(v.Str()))
+	case pcommon.ValueTypeBool:
+		if v.Bool() {
+			h.Write([]byte{1})
+		} else {
+			h.Write([]byte{0})
+		}
+	case pcommon.ValueTypeDouble:
+		buf := make([]byte, 8)
+		binary.LittleEndian.PutUint64(buf, math.Float64bits(v.Double()))
+		h.Write(buf)
+	case pcommon.ValueTypeInt:
+		buf := make([]byte, 8)
+		binary.LittleEndian.PutUint64(buf, uint64(v.Int()))
+		h.Write(buf)
+	case pcommon.ValueTypeBytes:
+		h.Write(v.Bytes().AsRaw())
+	case pcommon.ValueTypeMap:
+		mapHash(h, v.Map())
+	case pcommon.ValueTypeSlice:
+		sliceHash(h, v.Slice())
+	}
+}
+
+func sliceHash(h hash.Hash, s pcommon.Slice) {
+	for i := 0; i < s.Len(); i++ {
+		valueHash(h, s.At(i))
+	}
+}
```
I just came across https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/pdatautil/hash.go, which does a lot of this already (but using xxhash). Perhaps we should extend and use that?

I think the main thing we would be missing from pdatautil is the ability to keep state while hashing multiple values -- timestamp & attributes in our case. We could alternatively key on (timestamp, hash(attributes)).

Lastly, we could alternatively build on https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/exp/metrics/identity (specifically "OfStream"). But again, that doesn't include timestamp...

andrzej-stencel replied:
Thanks for finding pdatautil, I looked briefly for something like this but didn't see it. Definitely worth looking into it.

andrzej-stencel replied:
Due to a change in the implementation, the whole hashing part is now  removed.

axw reviewed (approved):
LGTM. I'm mildly concerned about using hash/fnv, particularly 32-bit, without any chaining for handling collisions. There's some existing code we can build on, but I think it would be reasonable to defer that to a followup.

andrzej-stencel committed:
fix lint

felixbarny commented on exporter/elasticsearchexporter/model_test.go:
```diff
@@ -22,6 +26,23 @@ var expectedSpanBody = `{"@timestamp":"2023-04-19T03:04:05.000000006Z","Attribut
 var expectedLogBody = `{"@timestamp":"2023-04-19T03:04:05.000000006Z","Attributes.log-attr1":"value1","Body":"log-body","Resource.key1":"value1","Scope.name":"","Scope.version":"","SeverityNumber":0,"TraceFlags":0}`
 
+var expectedMetricsEncoded = `{"@timestamp":"2024-06-12T10:20:16.419290690Z","cpu":"cpu0","host":{"name":"my-host"},"os":{"type":"linux"},"state":"idle","system":{"cpu":{"time":440.23}}}
+{"@timestamp":"2024-06-12T10:20:16.419290690Z","cpu":"cpu0","host":{"name":"my-host"},"os":{"type":"linux"},"state":"interrupt","system":{"cpu":{"time":0}}}
```
```suggestion
{"@timestamp":"2024-06-12T10:20:16.419290690Z","cpu":"cpu0","host":{"name":"my-host"},"os":{"type":"linux"},"state":"interrupt","system":{"cpu":{"time":0.0}}}
```

Let's try to force a fractional to be present for floating point numbers, so that ES can correctly do dynamic mappings.

andrzej-stencel replied:
Perhaps we can do this in a follow-up PR?

felixbarny replied:
sounds good

andrzej-stencel committed:
Merge remote-tracking branch 'upstream/main' into add-metrics-to-elasticsearch-exporter

andrzej-stencel committed:
Merge remote-tracking branch 'upstream/main' into add-metrics-to-elasticsearch-exporter

andrzej-stencel committed:
make generate

carsonip commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -163,6 +169,125 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
+			hash := metricHash(dp.Timestamp(), dp.Attributes())
+			doc, docExists := docs[hash]
+			if !docExists {
+				doc = baseDoc.Clone()
```
IIUC `Clone` is not a deep clone, and dedup is potentially going to mutate the field. I feel more comfortable with a deep clone. In any case, can you add a test where in resource attributes and dp attributes, there are duplicate fields, as well as renamed fields (see Dedup logic), and ensure that the output is still expected?

carsonip commented on exporter/elasticsearchexporter/model.go:
```diff
@@ -163,6 +169,125 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
+			dp := dps.At(j)
+
+			hash := metricHash(dp.Timestamp(), dp.Attributes())
+			doc, docExists := docs[hash]
```
non-blocker: Are we worried about hash collisions? In apm-data it doesn't just compare hashes, but compares string-based map keys, which in go implementation should first compare hash then actual bytes (see [code](https://github.com/elastic/apm-data/blob/main/input/otlp/metrics.go#L378)). However, at the same time, apm-data is vulnerable to value->string collisions.

andrzej-stencel replied:
Oh this whole hashing detail will actually be removed in a follow up PR ([draft here](https://github.com/andrzej-stencel/opentelemetry-collector-contrib/pull/1)) due to a change in how data is routed. Sorry for the trouble, probably no need reviewing this in light of the upcoming change.

andrzej-stencel commented on .chloggen/add-metrics-to-elasticsearch-exporter.yaml:
```diff
@@ -0,0 +1,27 @@
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []
```
The feature is available to users, so yes, it is relevant. `user` is the default value for `change_logs`, so no need to specify it explicitly.

carsonip replied:
TIL!

andrzej-stencel committed:
docs: change "events" to "metrics"

Co-authored-by: Carson Ip <carsonip@users.noreply.github.com>

andrzej-stencel committed:
fix test name

Co-authored-by: Carson Ip <carsonip@users.noreply.github.com>

andrzej-stencel committed:
Merge branch 'main' into add-metrics-to-elasticsearch-exporter
